lib/src/sql/regex.rs

```diff
@@ -4,59 +4,120 @@ use nom::bytes::complete::escaped;
 use std::str;
+use std::str::FromStr;
 
 pub(crate) const TOKEN: &str = "$surrealdb::private::sql::Regex";
 
-#[derive(Clone, Debug, Default, Eq, Ord, PartialEq, PartialOrd, Deserialize, Hash)]
-pub struct Regex(pub(super) String);
+#[derive(Clone)]
+pub struct Regex(pub(super) regex::Regex);
 
-impl From<&str> for Regex {
-	fn from(r: &str) -> Self {
-		Self(r.replace("\\/", "/"))
+impl Regex {
+	// Deref would expose `regex::Regex::as_str` which wouldn't have the '/' delimiters.
+	pub fn regex(&self) -> &regex::Regex {
+		&self.0
 	}
 }
 
-impl Deref for Regex {
-	type Target = String;
-	fn deref(&self) -> &Self::Target {
-		&self.0
+impl FromStr for Regex {
+	type Err = <regex::Regex as FromStr>::Err;
+
+	fn from_str(s: &str) -> Result<Self, Self::Err> {
+		regex::Regex::new(&s.replace("\\/", "/")).map(Self)
 	}
 }
 
-impl fmt::Display for Regex {
-	fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
-		write!(f, "/{}/", &self.0)
+impl PartialEq for Regex {
+	fn eq(&self, other: &Self) -> bool {
+		self.0.as_str().eq(other.0.as_str())
 	}
 }
 
-impl Regex {
-	pub fn regex(&self) -> Option<regex::Regex> {
-		regex::Regex::new(&self.0).ok()
+impl Eq for Regex {}
+
+impl Ord for Regex {
+	fn cmp(&self, other: &Self) -> Ordering {
+		self.0.as_str().cmp(other.0.as_str())
+	}
+}
+
+impl PartialOrd for Regex {
+	fn partial_cmp(&self, other: &Self) -> Option<Ordering> {
+		Some(self.cmp(other))
+	}
+}
+
+impl Hash for Regex {
+	fn hash<H: Hasher>(&self, state: &mut H) {
+		self.0.as_str().hash(state);
+	}
+}
+
+impl Debug for Regex {
+	fn fmt(&self, f: &mut Formatter) -> fmt::Result {
+		Display::fmt(self, f)
+	}
+}
+
+impl Display for Regex {
+	fn fmt(&self, f: &mut Formatter) -> fmt::Result {
+		write!(f, "/{}/", &self.0)
 	}
 }
 
 impl Serialize for Regex {
 	fn serialize<S>(&self, serializer: S) -> Result<S::Ok, S::Error>
 	where
-		S: serde::Serializer,
+		S: Serializer,
 	{
 		if is_internal_serialization() {
-			serializer.serialize_newtype_struct(TOKEN, &self.0)
+			serializer.serialize_newtype_struct(TOKEN, self.0.as_str())
 		} else {
 			serializer.serialize_none()
 		}
 	}
 }
 
+pub struct RegexVisitor;
+
+impl<'de> Visitor<'de> for RegexVisitor {
+	type Value = Regex;
+
+	fn expecting(&self, formatter: &mut fmt::Formatter) -> fmt::Result {
+		formatter.write_str("a regex str")
+	}
+
+	fn visit_str<E>(self, value: &str) -> Result<Self::Value, E>
+	where
+		E: de::Error,
+	{
+		Regex::from_str(value).map_err(|_| de::Error::custom("invalid regex"))
+	}
+}
+
+impl<'de> Deserialize<'de> for Regex {
+	fn deserialize<D>(deserializer: D) -> Result<Self, D::Error>
+	where
+		D: Deserializer<'de>,
+	{
+		deserializer.deserialize_str(RegexVisitor)
+	}
+}
+
 pub fn regex(i: &str) -> IResult<&str, Regex> {
 	let (i, _) = char('/')(i)?;
 	let (i, v) = escaped(is_not("\\/"), '\\', anychar)(i)?;
 	let (i, _) = char('/')(i)?;
-	Ok((i, Regex::from(v)))
+	let regex = Regex::from_str(v).map_err(|_| nom::Err::Error(crate::sql::Error::Parser(v)))?;
```

Comment:
I know we already have `FromStr` in scope here but I think

```rust
let regex = v.parse().map_err(|_| nom::Err::Error(crate::sql::Error::Parser(v)))?;
```

might still be a tad readable.